File: cpp/prob002.cpp
#include <iostream>
#include <chrono>
#include "utils.h"

#define PROB_NUM    2

int compute()
{
    int fib[2] = {1,2};
    int sum = 2, n;

    while(n < 4000000){
        n = fib[0] + fib[1];
        fib[0] = fib[1];
        fib[1] = n;
        if(n%2 == 0)
            sum += n;
    }
    return sum;
}


int test_answer(std::string prob_answer) {
    std::string answer;
    if(!get_answer(PROB_NUM, &answer)) {
        std::cout << "file not found" << std::endl;
        return 1;
    }

    assert(answer.compare(prob_answer) == 0);
    return 0;
}

int main(int argc, char const *argv[])
{
    auto t1 = std::chrono::high_resolution_clock::now();
    int64_t result = compute();
    auto t2 = std::chrono::high_resolution_clock::now();

    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
    std::cout << result << "\tin " << duration << " us\n";

    // do testing
    test_answer(std::to_string(result));
    return 0;
}

File: rust/src/bin/prob002.rs
extern crate euler;

fn compute() -> u64{
    (1..).map(|n| euler::fibonacci(n))
        .take_while(|&x| x < 4000000)
        .filter(|&x| x%2 == 0)
        .sum()
}

fn main() {
    let now = std::time::Instant::now();
    let result = compute();
    println!("{:?}\tin {}us", result, now.elapsed().as_micros());
}


#[cfg(test)]
mod tests {
    use super::*;
    const PROB_NUM: usize = 2;
    
    #[test]
    fn test_prob002() {
        let result = compute().to_string();
        euler::test_answer(PROB_NUM, result);
    }
}